File: 3639-zero-array-transformation-i/3639-zero-array-transformation-i.cpp
class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        vector<int> diff(n + 1);
        for (auto query : queries) {
            int l = query[0], r = query[1];
            diff[l] += 1;
            diff[r + 1] -= 1;
        }

        int sum = 0;
        for (int i = 0; i < n; i++) {
            sum += diff[i];
            if (sum < nums[i])
                return false;
        }

        return true;
    }
};

File: 3639-zero-array-transformation-i/3639-zero-array-transformation-i.py
class Solution:
    def isZeroArray(self, nums: List[int], queries: List[List[int]]) -> bool:
        n = len(nums)
        arr = [0] * (n + 1)

        for query in queries:
            x, y = query[0], query[1]
            arr[x] += -1
            arr[y + 1] += 1

        for i in range(1, n + 1):
            arr[i] += arr[i - 1]

        print(arr)

        for i in range(n):
            if nums[i] + arr[i] > 0:
                return False
        return True
